src/hotspot/cpu/riscv/nativeInst_riscv.cpp

```diff
@@ -89,22 +89,47 @@ void NativeCall::print() {
   tty->print_cr(PTR_FORMAT ": auipc,ld,jalr x1, offset/reg, ", p2i(instruction_address()));
 }
 
+void NativeCall::optimize_call(address dest, bool mt_safe) {
+  // Skip over auipc + ld
+  address jal_pc = instruction_address() + 2 * NativeInstruction::instruction_size;
+  uint32_t *jal_pos = (uint32_t *)jal_pc;
+  // If reachable use JAL
+  if (Assembler::reachable_from_branch_at(jal_pc, dest)) {
+    int64_t distance = dest - jal_pc;
+    uint32_t new_jal = Assembler::encode_jal(ra, distance);
+    Atomic::store(jal_pos, new_jal);
+  } else if (!MacroAssembler::is_jalr_at(jal_pc)) { // The jalr is always identical: jalr ra, 0(t1)
+    uint32_t new_jal = Assembler::encode_jalr(ra, t1, 0);
+    Atomic::store(jal_pos, new_jal);
+  } else {
+    // No change to instruction stream
+    return;
+  }
+  // We changed instruction stream
+  if (mt_safe) {
+    OrderAccess::release();
+    ICache::invalidate_range(jal_pc, NativeInstruction::instruction_size);
```

Comment:
should `jal_pc` be `instruction_address()`?

Reply:
fixed